Inability to Iterate Over Entries in SIFS-Backed Multimap

### Describe the bug

Iterating over keyset of SIFS is problematic

### Expected behavior

It should be possible to iterate over entries in a SIFS-backed multimap directly. However, this functionality fails when dealing with a large number of entries (approximately 30 million). As a workaround, it should still be possible to iterate over the keyset and fetch each entry individually.

The issue becomes apparent after extended usage or when the multimap grows beyond a certain threshold, leading to the following problems:

Iteration over the keyset becomes unreliable.


Using the work around, to fetch each segment, and not full index in one step:
```
 try {
    log.info("Clone segment: {}", segment);
    org.infinispan.commons.util.IntSet intSet = IntSets.immutableSet(segment);
    CloseableIteratorSet<String> keySet = remoteEventCacheOffline.keySet(intSet);
    Set<String> keys = keySet.stream() //
            .collect(Collectors.toSet());
    return keys;
}catch (Exception ex) {
    log.info("IllegalStateException add to retry set: {} : {}", segment, ex.getMessage());
    retrySet.add(segment);
}
```

### Actual behavior

We get a exception, interesting is that: "Segment 116" - When actually segment 7 was exported with the above code.

```
2024-12-02 16:20:11,829 WARN  [org.infinispan.HOTROD] (HotRod-client-async-pool-4-71) ISPN004098: Closing connection [id: 0x468fb385, L:/49.12.*2.*:33238 - R:49.12.*.*/49.12.*.*:11222] due to transport error: java.lang.IllegalStateException: Segment 116 already completed
at deployment.inbound-ear-5.9.2138.ear//org.infinispan.client.hotrod.impl.iteration.SegmentKeyTracker.track(SegmentKeyTracker.java:66)
at deployment.inbound-ear-5.9.2138.ear//org.infinispan.client.hotrod.impl.operations.IterationNextOperation.acceptResponse(IterationNextOperation.java:120)
at deployment.inbound-ear-5.9.2138.ear//org.infinispan.client.hotrod.impl.transport.netty.HeaderDecoder.decode(HeaderDecoder.java:153)
at deployment.inbound-ear-5.9.2138.ear//org.infinispan.client.hotrod.impl.transport.netty.HintedReplayingDecoder.callDecode(HintedReplayingDecoder.java:94)
at deployment.inbound-ear-5.9.2138.ear//io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at deployment.inbound-ear-5.9.2138.ear//io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:801)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501)
at deployment.inbound-ear-5.9.2138.ear//io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399)
at deployment.inbound-ear-5.9.2138.ear//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
at deployment.inbound-ear-5.9.2138.ear//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [java.base:]
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [java.base:]
at java.base/java.lang.Thread.run(Thread.java:833) [java.base:
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Infinispan version or git rev

quay.io/infinispan/server:15.0.10.Final-1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

One full loop over all segment, with one retry of the faile ones:
`IllegalStateException add to retry set: 2 : Segment 0 already completed
IllegalStateException add to retry set: 7 : Segment 116 already completed
IllegalStateException add to retry set: 11 : Segment 6 already completed
IllegalStateException add to retry set: 12 : Segment 9 already completed
IllegalStateException add to retry set: 13 : Segment 9 already completed
IllegalStateException add to retry set: 24 : Segment 19 already completed
IllegalStateException add to retry set: 25 : Segment 22 already completed
IllegalStateException add to retry set: 31 : Segment 29 already completed
IllegalStateException add to retry set: 32 : Segment 30 already completed
IllegalStateException add to retry set: 33 : Segment 30 already completed
IllegalStateException add to retry set: 36 : Segment 34 already completed
IllegalStateException add to retry set: 37 : Segment 35 already completed
IllegalStateException add to retry set: 40 : org.infinispan.commons.CacheException: Error reading header from 6096:10141981 | 0
IllegalStateException add to retry set: 41 : Segment 39 already completed
IllegalStateException add to retry set: 44 : Segment 42 already completed
IllegalStateException add to retry set: 45 : Segment 43 already completed
IllegalStateException add to retry set: 48 : Segment 46 already completed
IllegalStateException add to retry set: 49 : Segment 47 already completed
IllegalStateException add to retry set: 53 : Segment 51 already completed
IllegalStateException add to retry set: 56 : Segment 54 already completed
IllegalStateException add to retry set: 57 : Segment 55 already completed
IllegalStateException add to retry set: 61 : Segment 59 already completed
IllegalStateException add to retry set: 66 : Segment 21 already completed
IllegalStateException add to retry set: 67 : Segment 64 already completed
IllegalStateException add to retry set: 72 : Segment 70 already completed
IllegalStateException add to retry set: 76 : Segment 74 already completed
IllegalStateException add to retry set: 77 : Segment 75 already completed
IllegalStateException add to retry set: 78 : Segment 75 already completed
IllegalStateException add to retry set: 79 : Segment 75 already completed
IllegalStateException add to retry set: 82 : Segment 80 already completed
IllegalStateException add to retry set: 83 : Segment 73 already completed
IllegalStateException add to retry set: 100 : Segment 81 already completed
IllegalStateException add to retry set: 101 : Segment 98 already completed
IllegalStateException add to retry set: 104 : Segment 99 already completed
IllegalStateException add to retry set: 105 : Segment 102 already completed
IllegalStateException add to retry set: 110 : Segment 107 already completed
IllegalStateException add to retry set: 111 : Segment 108 already completed
IllegalStateException add to retry set: 118 : Segment 112 already completed
IllegalStateException add to retry set: 119 : Segment 113 already completed
IllegalStateException add to retry set: 123 : Segment 120 already completed
IllegalStateException add to retry set: 131 : Segment 129 already completed
IllegalStateException add to retry set: 132 : Segment 130 already completed
IllegalStateException add to retry set: 135 : Segment 193 already completed
IllegalStateException add to retry set: 136 : Segment 134 already completed
IllegalStateException add to retry set: 153 : Segment 151 already completed
IllegalStateException add to retry set: 154 : Segment 152 already completed
IllegalStateException add to retry set: 166 : Segment 54 already completed
IllegalStateException add to retry set: 167 : Segment 165 already completed
IllegalStateException add to retry set: 180 : org.infinispan.commons.CacheException: Error reading header from 21513:6333329 | 0
IllegalStateException add to retry set: 181 : Segment 61 already completed
IllegalStateException add to retry set: 195 : Segment 193 already completed
IllegalStateException add to retry set: 196 : Segment 194 already completed
IllegalStateException add to retry set: 197 : Segment 194 already completed
IllegalStateException add to retry set: 202 : Segment 122 already completed
IllegalStateException add to retry set: 206 : Segment 203 already completed
IllegalStateException add to retry set: 209 : Segment 207 already completed
IllegalStateException add to retry set: 210 : Segment 208 already completed
IllegalStateException add to retry set: 214 : Segment 205 already completed
IllegalStateException add to retry set: 215 : Segment 212 already completed
IllegalStateException add to retry set: 218 : Segment 213 already completed
IllegalStateException add to retry set: 219 : Segment 217 already completed
IllegalStateException add to retry set: 224 : Segment 201 already completed
IllegalStateException add to retry set: 225 : Segment 222 already completed
IllegalStateException add to retry set: 231 : Segment 228 already completed
IllegalStateException add to retry set: 232 : Segment 229 already completed
IllegalStateException add to retry set: 238 : Segment 235 already completed
IllegalStateException add to retry set: 239 : Segment 236 already completed
IllegalStateException add to retry set: 243 : Segment 241 already completed
IllegalStateException add to retry set: 247 : Segment 245 already completed
IllegalStateException add to retry set: 248 : Segment 246 already completed
IllegalStateException add to retry set: 253 : Segment 251 already completed
IllegalStateException add to retry set: 254 : Segment 252 already completed
IllegalStateException add to retry set: 2 : Segment 252 already completed
IllegalStateException add to retry set: 131 : Segment 252 already completed
IllegalStateException add to retry set: 132 : Segment 252 already completed
IllegalStateException add to retry set: 136 : Segment 135 already completed
IllegalStateException add to retry set: 12 : Segment 7 already completed
IllegalStateException add to retry set: 25 : Segment 192 already completed
IllegalStateException add to retry set: 154 : Segment 153 already completed
IllegalStateException add to retry set: 32 : Segment 31 already completed
IllegalStateException add to retry set: 36 : Segment 33 already completed
IllegalStateException add to retry set: 167 : Segment 166 already completed
IllegalStateException add to retry set: 40 : org.infinispan.commons.CacheException: Error reading header from 6096:10141981 | 0
IllegalStateException add to retry set: 44 : Segment 41 already completed
IllegalStateException add to retry set: 48 : Segment 45 already completed
IllegalStateException add to retry set: 180 : Segment 49 already completed
IllegalStateException add to retry set: 181 : Segment 49 already completed
IllegalStateException add to retry set: 57 : Segment 56 already completed
IllegalStateException add to retry set: 67 : Segment 13 already completed
IllegalStateException add to retry set: 196 : Segment 195 already completed
IllegalStateException add to retry set: 72 : Segment 169 already completed
IllegalStateException add to retry set: 77 : Segment 53 already completed
IllegalStateException add to retry set: 206 : Segment 76 already completed
IllegalStateException add to retry set: 79 : Segment 76 already completed
IllegalStateException add to retry set: 215 : Segment 82 already completed
IllegalStateException add to retry set: 219 : Segment 214 already completed
IllegalStateException add to retry set: 225 : Segment 83 already completed
IllegalStateException add to retry set: 100 : Segment 83 already completed
IllegalStateException add to retry set: 231 : Segment 224 already completed
IllegalStateException add to retry set: 232 : Segment 101 already completed
IllegalStateException add to retry set: 110 : Segment 104 already completed
IllegalStateException add to retry set: 111 : Segment 105 already completed
IllegalStateException add to retry set: 239 : Segment 105 already completed
IllegalStateException add to retry set: 119 : Segment 238 already completed
IllegalStateException add to retry set: 248 : Segment 218 already completed
IllegalStateException add to retry set: 254 : Segment 253 already completed`

I think segment 40 is really broken?
from the server for segment 40:
`Caused by: java.lang.IllegalStateException: Error reading header from 6096:10141981 | 0
at org.infinispan.persistence.sifs.IndexNode$LeafNode.getHeaderAndKey(IndexNode.java:1126)
at org.infinispan.persistence.sifs.IndexNode.recursiveNode(IndexNode.java:1294)
at org.infinispan.persistence.sifs.IndexNode.recursiveNode(IndexNode.java:1273)
at org.infinispan.persistence.sifs.IndexNode.recursiveNode(IndexNode.java:1273)
at org.infinispan.persistence.sifs.IndexNode.recursiveNode(IndexNode.java:1273)
at org.infinispan.persistence.sifs.IndexNode.lambda$publish$1(IndexNode.java:1250)
at org.infinispan.reactive.FlowableCreate$BaseEmitter.attemptSubscribe(FlowableCreate.java:338)
at org.infinispan.reactive.FlowableCreate$BaseEmitter.request(FlowableCreate.java:367)
at io.reactivex.rxjava3.internal.operators.flowable.FlowableDoFinally$DoFinallySubscriber.request(FlowableDoFinally.java:108)
at io.reactivex.rxjava3.internal.subscriptions.SubscriptionArbiter.request(SubscriptionArbiter.java:123)
at io.reactivex.rxjava3.internal.operators.flowable.FlowableConcatMap$ConcatMapImmediate.request(FlowableConcatMap.java:217)
at io.reactivex.rxjava3.internal.subscribers.BasicFuseableSubscriber.request(BasicFuseableSubscriber.java:153)
at io.reactivex.rxjava3.internal.subscriptions.SubscriptionArbiter.request(SubscriptionArbiter.java:123)
at io.reactivex.rxjava3.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber$Request.run(FlowableSubscribeOn.java:156)
at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:324)
at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker.runEager(ExecutorScheduler.java:289)
at io.reactivex.rxjava3.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:250)
at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
... 1 more`

--- a/core/src/main/java/org/infinispan/persistence/sifs/Compactor.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/Compactor.java
@@ -103,6 +103,19 @@ class Compactor {
       recordFreeSpace(getStats(file, -1, -1), file, size);
    }
 
+   // Adds the free size if the file is present or is a log file (which may not yet be present)
+   public void freeIfPresent(int file, int size) {
+      Stats stats;
+      // Log file check is first since a file goes from being a log file to being in the stats
+      if (fileProvider.isLogFile(file)) {
+         // Technically the file could be compacted and removed between these calls but at worst it just has
+         // a file with a stat that is not completed and is effectively ignored
+         free(file, size);
+      } else if ((stats = fileStats.get(file)) != null) {
+         recordFreeSpace(stats, file, size);
+      }
+   }
+
    public void completeFile(int file, int currentSize, long nextExpirationTime) {
       completeFile(file, currentSize, nextExpirationTime, true);
    }

--- a/core/src/main/java/org/infinispan/persistence/sifs/FileProvider.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/FileProvider.java
@@ -50,6 +50,7 @@ public class FileProvider {
    private final Set<FileIterator> iterators = ConcurrentHashMap.newKeySet();
    private final String prefix;
    private final int maxFileSize;
+   private final boolean isIndex;
 
    private boolean canTryPmem = true;
 
@@ -67,12 +68,14 @@ public class FileProvider {
       ATTEMPT_PMEM = attemptPmem;
    }
 
-   public FileProvider(Path fileDirectory, int openFileLimit, String prefix, int maxFileSize) {
+   public FileProvider(Path fileDirectory, int openFileLimit, String prefix, int maxFileSize,
+                       boolean isIndex) {
       this.openFileLimit = openFileLimit;
       this.recordQueue = new ArrayBlockingQueue<>(openFileLimit);
       this.directoryFile = fileDirectory.toFile();
       this.prefix = prefix;
       this.maxFileSize = maxFileSize;
+      this.isIndex = isIndex;
       try {
          Files.createDirectories(fileDirectory);
       } catch (IOException e) {
@@ -236,7 +239,7 @@ public class FileProvider {
          if (create) {
             fileChannel = new FileOutputStream(file).getChannel();
          } else {
-            fileChannel = new RandomAccessFile(file, "rw").getChannel();
+            fileChannel = new RandomAccessFile(file, isIndex ? "rw" : "r").getChannel();
          }
       }
 

--- a/core/src/main/java/org/infinispan/persistence/sifs/Index.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/Index.java
@@ -143,7 +143,7 @@ class Index {
       this.minNodeSize = minNodeSize;
       this.maxNodeSize = maxNodeSize;
       this.sizePerSegment = new AtomicLongArray(cacheSegments);
-      this.indexFileProvider = new FileProvider(indexDir, maxOpenFiles, "index.", Integer.MAX_VALUE);
+      this.indexFileProvider = new FileProvider(indexDir, maxOpenFiles, "index.", Integer.MAX_VALUE, true);
       this.indexSizeFile = new File(indexDir.toFile(), "index-count");
 
       this.segments = new Segment[cacheSegments];

--- a/core/src/main/java/org/infinispan/persistence/sifs/Index.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/Index.java
@@ -703,9 +703,9 @@ class Index {
                                  // Now we free all entries in the index, this will include expired and removed entries
                                  // Removed doesn't currently update free stats per ISPN-15246 - so we remove those as well
                                  segment.root.publish((keyAndMetadataRecord, leafNode, fileProvider, timeService) -> {
-                                          compactor.free(leafNode.file, keyAndMetadataRecord.getHeader().totalLength());
+                                          compactor.freeIfPresent(leafNode.file, keyAndMetadataRecord.getHeader().totalLength());
                                           return null;
-                                       }).ignoreElements()
+                                       }, true).ignoreElements()
                                        .toCompletionStage(null)
                            )
                            .thenRun(segment::delete));
@@ -1103,7 +1103,7 @@ class Index {
                return leafNode.loadValue(keyAndMetadataRecord, fileProvider);
             }
             return keyAndMetadataRecord;
-         }).doFinally(() -> lock.unlockRead(stamp));
+         }, false).doFinally(() -> lock.unlockRead(stamp));
       } catch (Throwable t) {
          lock.unlockRead(stamp);
          throw t;

--- a/core/src/main/java/org/infinispan/persistence/sifs/IndexNode.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/IndexNode.java
@@ -1242,7 +1242,7 @@ class IndexNode {
             throws IOException, IndexNodeOutdatedException;
    }
 
-   <R> Flowable<R> publish(PublishFunction<R> publishFunction) {
+   <R> Flowable<R> publish(PublishFunction<R> publishFunction, boolean continueOnOutdated) {
       long currentTime = segment.getTimeService().wallClockTime();
 
       // Needs defer as we mutate the lastRetrievedKey so inner FlowableCreate can be subscribed to multiple times
@@ -1256,7 +1256,7 @@ class IndexNode {
                // Reset so we can loop
                done.set(false);
                recursiveNode(publishFunction, this, segment, lastRetrievedKey, emitter, currentTime,
-                     new ByRef.Boolean(false), done);
+                     new ByRef.Boolean(false), done, continueOnOutdated);
                // This handles two of the done cases - in which case we can't continue
                if (emitter.requested() == 0 || emitter.isCancelled()) {
                   return;
@@ -1268,7 +1268,7 @@ class IndexNode {
    }
 
    <R> void recursiveNode(PublishFunction<R> publishFunction, IndexNode node, Index.Segment segment, ByRef<byte[]> lastRetrievedKey, FlowableEmitter<R> emitter,
-         long currentTime, ByRef.Boolean foundData, ByRef.Boolean done) throws IOException {
+         long currentTime, ByRef.Boolean foundData, ByRef.Boolean done, boolean continueOnOutdated) throws IOException {
       Lock readLock = node.lock.readLock();
       readLock.lock();
       try {
@@ -1279,7 +1279,7 @@ class IndexNode {
             // Need to search all inner nodes starting from that point until we hit the last entry for the segment
             for (int i = point; i < node.innerNodes.length && !done.get(); ++i) {
                recursiveNode(publishFunction, node.innerNodes[i].getIndexNode(segment), segment, lastRetrievedKey, emitter,
-                     currentTime, foundData, done);
+                     currentTime, foundData, done, continueOnOutdated);
             }
          } else if (node.leafNodes != null) {
             int suggestedIteration;
@@ -1332,6 +1332,11 @@ class IndexNode {
                   }
                   previousKey = record.getKey();
                } catch (IndexNodeOutdatedException e) {
+                  if (continueOnOutdated) {
+                     // Continue to next indexed key instead of restarting from previous key
+                     log.tracef("Ignoring IndexNodeOutdatedException for segment %s iteration", segment.getId());
+                     continue;
+                  }
                   // Current key was outdated, we have to try from the previous entry we saw (note it is skipped)
                   if (previousKey != null) {
                      lastRetrievedKey.set(previousKey);

--- a/core/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java
+++ b/core/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java
@@ -231,7 +231,7 @@ public class NonBlockingSoftIndexFileStore<K, V> implements NonBlockingStore<K, 
       }
 
       fileProvider = new FileProvider(getDataLocation(), maxOpenDataFiles, PREFIX_LATEST,
-            configuration.maxFileSize());
+            configuration.maxFileSize(), false);
       compactor = new Compactor(ctx.getNonBlockingManager(), fileProvider, temporaryTable, marshaller, timeService,
             keyPartitioner, configuration.maxFileSize(), configuration.compactionThreshold(),
             blockingManager.asExecutor("sifs-compactor"));
@@ -259,7 +259,7 @@ public class NonBlockingSoftIndexFileStore<K, V> implements NonBlockingStore<K, 
          // we don't destroy the data on startup
          // get the old files
          FileProvider oldFileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit(), PREFIX_10_1,
-             configuration.maxFileSize());
+             configuration.maxFileSize(), false);
          // Don't even try to rebuild index or read files if we are allowed to directly purge on startup
          if (ctx.canStoreDirectlyPurgeOnStartup()) {
             try {
@@ -273,7 +273,7 @@ public class NonBlockingSoftIndexFileStore<K, V> implements NonBlockingStore<K, 
                throw PERSISTENCE.persistedDataMigrationUnsupportedVersion("< 11");
             }
             oldFileProvider = new FileProvider(getDataLocation(), configuration.openFilesLimit(), PREFIX_11_0,
-                  configuration.maxFileSize());
+                  configuration.maxFileSize(), false);
             if (oldFileProvider.hasFiles()) {
                try {
                   index.reset();

--- a/core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java
+++ b/core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java
@@ -539,7 +539,9 @@ public class EvictionWithConcurrentOperationsTest extends SingleCacheManagerTest
    }
 
    @AfterClass(alwaysRun = true)
-   protected void clearTempDir() {
+   @Override
+   protected void destroyAfterClass() {
+      super.destroyAfterClass();
       Util.recursiveFileRemove(persistentLocation);
    }
 

--- a/core/src/test/java/org/infinispan/expiration/impl/ExpirationFileStoreListenerFunctionalTest.java
+++ b/core/src/test/java/org/infinispan/expiration/impl/ExpirationFileStoreListenerFunctionalTest.java
@@ -58,7 +58,9 @@ public class ExpirationFileStoreListenerFunctionalTest extends ExpirationStoreLi
    }
 
    @AfterClass(alwaysRun = true)
-   protected void clearTempDir() {
+   @Override
+   protected void destroyAfterClass() {
+      super.destroyAfterClass();
       Util.recursiveFileRemove(Testing.tmpDirectory(this.getClass()));
    }
 }

--- a/core/src/test/java/org/infinispan/persistence/sifs/FileProviderTest.java
+++ b/core/src/test/java/org/infinispan/persistence/sifs/FileProviderTest.java
@@ -1,0 +1,104 @@
+package org.infinispan.persistence.sifs;
+
+import static org.testng.AssertJUnit.assertNotNull;
+import static org.testng.AssertJUnit.assertNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
+import org.infinispan.commons.util.Util;
+import org.infinispan.testing.Testing;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "persistence.sifs.FileProviderTest")
+public class FileProviderTest {
+
+   private String tmpDirectory;
+
+   @BeforeMethod
+   protected void setUpTempDir() {
+      tmpDirectory = Testing.tmpDirectory(getClass());
+   }
+
+   @AfterMethod
+   protected void clearTempDir() {
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   public void testGetFileReadsEmptyFileWhenIsIndexFalse() throws IOException {
+      Path dataPath = Path.of(tmpDirectory, "data");
+      FileProvider fileProvider = new FileProvider(dataPath, 10, "test-", 1000, false);
+
+      try {
+         // Create an empty file to simulate the scenario
+         File emptyFile = fileProvider.newFile(1);
+         emptyFile.createNewFile();
+
+         // When isIndex is false, opening an existing empty file with "r" mode should succeed
+         // (read-only mode can open existing empty files)
+         FileProvider.Handle handle = fileProvider.getFile(1);
+         assertNotNull("Expected getFile to succeed for existing empty file when isIndex is false", handle);
+
+         if (handle != null) {
+            handle.close();
+         }
+      } finally {
+         fileProvider.stop();
+      }
+   }
+
+   public void testGetFileSucceedsWhenIsIndexTrue() throws IOException {
+      Path indexPath = Path.of(tmpDirectory, "index");
+      FileProvider fileProvider = new FileProvider(indexPath, 10, "test-", 1000, true);
+
+      try {
+         // Create an empty file to simulate the scenario
+         File emptyFile = fileProvider.newFile(1);
+         emptyFile.createNewFile();
+
+         // When isIndex is true, opening with "rw" mode should succeed even for empty files
+         FileProvider.Handle handle = fileProvider.getFile(1);
+         assertNotNull("Expected getFile to succeed for empty file when isIndex is true", handle);
+
+         if (handle != null) {
+            handle.close();
+         }
+      } finally {
+         fileProvider.stop();
+      }
+   }
+
+   public void testGetFileNonExistentFileWhenIsIndexFalse() throws IOException {
+      Path dataPath = Path.of(tmpDirectory, "data");
+      FileProvider fileProvider = new FileProvider(dataPath, 10, "test-", 1000, false);
+
+      try {
+         // When isIndex is false, opening a non-existent file with "r" mode should fail
+         // getFile catches FileNotFoundException and returns null
+         FileProvider.Handle handle = fileProvider.getFile(999);
+         assertNull("Expected getFile to return null for non-existent file when isIndex is false", handle);
+      } finally {
+         fileProvider.stop();
+      }
+   }
+
+   public void testGetFileNonExistentFileWhenIsIndexTrue() throws IOException {
+      Path indexPath = Path.of(tmpDirectory, "index");
+      FileProvider fileProvider = new FileProvider(indexPath, 10, "test-", 1000, true);
+
+      try {
+         // When isIndex is true, opening with "rw" mode should create the file if it doesn't exist
+         FileProvider.Handle handle = fileProvider.getFile(999);
+         assertNotNull("Expected getFile to create and open file when isIndex is true", handle);
+
+         if (handle != null) {
+            handle.close();
+         }
+      } finally {
+         fileProvider.stop();
+      }
+   }
+}

--- a/core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreSegmentRemovalTest.java
+++ b/core/src/test/java/org/infinispan/persistence/sifs/SoftIndexFileStoreSegmentRemovalTest.java
@@ -1,0 +1,192 @@
+package org.infinispan.persistence.sifs;
+
+import static org.testng.AssertJUnit.assertEquals;
+import static org.testng.AssertJUnit.assertNull;
+import static org.testng.AssertJUnit.assertTrue;
+import static org.testng.AssertJUnit.fail;
+
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.util.IntSets;
+import org.infinispan.commons.util.Util;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.global.GlobalConfigurationBuilder;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.test.SingleCacheManagerTest;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.TestCacheManagerFactory;
+import org.infinispan.testing.Testing;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import io.reactivex.rxjava3.core.Flowable;
+
+/**
+ * Tests for SoftIndexFileStore segment removal and compaction scenarios
+ */
+@Test(groups = "functional", testName = "persistence.sifs.SoftIndexFileStoreSegmentRemovalTest")
+public class SoftIndexFileStoreSegmentRemovalTest extends SingleCacheManagerTest {
+   private static final Log log = Log.getLog(SoftIndexFileStoreSegmentRemovalTest.class);
+
+   private String tmpDirectory;
+
+   @BeforeClass(alwaysRun = true)
+   @Override
+   protected void createBeforeClass() throws Exception {
+      tmpDirectory = Testing.tmpDirectory(getClass());
+      super.createBeforeClass();
+   }
+
+   @AfterClass(alwaysRun = true)
+   @Override
+   protected void destroyAfterClass() {
+      super.destroyAfterClass();
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   @Override
+   protected EmbeddedCacheManager createCacheManager() throws Exception {
+      GlobalConfigurationBuilder global = new GlobalConfigurationBuilder();
+      global.globalState().enable().persistentLocation(Testing.tmpDirectory(this.getClass()));
+
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.clustering().hash().numSegments(2);
+      builder.persistence()
+            .addSoftIndexFileStore()
+            .dataLocation(Paths.get(tmpDirectory, "data").toString())
+            .indexLocation(Paths.get(tmpDirectory, "index").toString())
+            .maxFileSize(1000);
+
+      return TestCacheManagerFactory.newDefaultCacheManager(true, global, builder);
+   }
+
+   /**
+    * Tests concurrent execution of removeSegments and compaction.
+    * This is a regression test for scenarios where:
+    * 1. addSegments is called to initialize segments
+    * 2. Data is written to a segment
+    * 3. removeSegments is invoked, which asynchronously cleans old index information
+    * 4. While that cleanup is blocked/delayed, compaction is triggered
+    * 5. After compaction completes, the cleanup proceeds
+    * 6. Both operations complete successfully without corruption
+    */
+   public void testRemoveSegmentsDuringCompaction() throws Exception {
+      Cache<String, String> cache = cacheManager.getCache();
+
+      // Write enough values into segment 0 to force the log appender to open another file
+      // maxFileSize is 1000 bytes, so we need to write enough data to exceed that
+      // Use larger values with unique content to prevent early compaction
+      StringBuilder largeValue = new StringBuilder(200);
+      for (int j = 0; j < 200; j++) {
+         largeValue.append('x');
+      }
+
+      for (int i = 0; i < 20; i++) {
+         String key = "key-" + i;
+         // Add unique nanoTime to prevent compaction from deduplicating
+         String value = largeValue + "-" + i + "-" + System.nanoTime();
+         cache.put(key, value);
+      }
+
+      NonBlockingSoftIndexFileStore store = TestingUtil.getFirstStore(cache);
+      Compactor compactor = TestingUtil.extractField(store, "compactor");
+      Index index = TestingUtil.extractField(store, "index");
+
+      // Verify that multiple files were created
+      int fileCount = compactor.getFiles().size();
+      assertTrue("Should have multiple files for compaction, but had " + fileCount, fileCount > 1);
+
+      CountDownLatch cleanupStarted = new CountDownLatch(1);
+      CountDownLatch compactionComplete = new CountDownLatch(1);
+      AtomicBoolean firstTaskSeen = new AtomicBoolean(false);
+
+      // Get the original executor
+      Executor originalExecutor = TestingUtil.extractField(index, "executor");
+
+      // Create a delaying executor that blocks the first task
+      Executor delayingExecutor = command -> {
+         if (firstTaskSeen.compareAndSet(false, true)) {
+            // This is the first task (from removeSegments)
+            originalExecutor.execute(() -> {
+               // Signal that cleanup has started
+               log.trace("Delaying executor: first task started, signaling latch");
+               cleanupStarted.countDown();
+               // Wait for compaction to complete
+               try {
+                  log.trace("Delaying executor: waiting for compaction to complete...");
+                  boolean completed = compactionComplete.await(10, TimeUnit.SECONDS);
+                  log.tracef("Delaying executor: compaction complete signal received: %s", completed);
+                  assertTrue("Compaction should complete before cleanup proceeds", completed);
+               } catch (InterruptedException e) {
+                  throw new RuntimeException(e);
+               }
+               // Now execute the actual task
+               command.run();
+            });
+         } else {
+            // All subsequent tasks run normally
+            originalExecutor.execute(command);
+         }
+      };
+
+      // Inject the delaying executor
+      TestingUtil.replaceField(delayingExecutor, "executor", index, Index.class);
+
+      try {
+         // Start removeSegments in background - the executor invokes the cleanup part
+         store.removeSegments(IntSets.immutableSet(0)).toCompletableFuture().get(10, TimeUnit.SECONDS);
+
+         // Wait for cleanup to start
+         assertTrue("Cleanup should have started", cleanupStarted.await(10, TimeUnit.SECONDS));
+
+         // Now invoke compaction while the segment cleanup is blocked
+         try {
+            log.trace("Starting compaction...");
+            compactor.forceCompactionForAllNonLogFiles().toCompletableFuture().get(10, TimeUnit.SECONDS);
+            log.trace("Compaction completed successfully");
+         } catch (Exception e) {
+            log.tracef(e, "Compaction failed: %s", e.getMessage());
+         } finally {
+            // Signal compaction is complete (even if it failed)
+            compactionComplete.countDown();
+            log.trace("Signaled compaction complete");
+         }
+
+         // Extract and wait for the removeSegmentsStage to complete
+         // This will throw any exception that occurred during the cleanup task
+         CompletionStage<Void> removeSegmentsStage = TestingUtil.extractField(index, "removeSegmentsStage");
+         removeSegmentsStage.toCompletableFuture().get(10, TimeUnit.SECONDS);
+
+         // Verify file stats are the same after removeSegments
+         Map<Integer, Compactor.Stats> fileStats = compactor.getFileStats();
+         // Note that total of -1 is alright, that means it is a non completed file
+         if (fileStats.values().stream().anyMatch(s -> s.getTotal() == 0)) {
+            fail("At least one stat shows no total: " + fileStats);
+         }
+
+         // Verify the store is still in a consistent state - entries should be removed
+         assertNull(TestingUtil.join(store.load(0, "key-0")));
+         assertNull(TestingUtil.join(store.load(0, "key-10")));
+         assertNull(TestingUtil.join(store.load(0, "key-19")));
+
+         // Verify that iteration doesn't return any of the removed values
+         AtomicInteger iterationCount = new AtomicInteger(0);
+         Flowable.fromPublisher(
+               store.publishEntries(IntSets.immutableSet(0), null, true))
+               .blockingForEach(entry -> iterationCount.incrementAndGet(), 128);
+         assertEquals("Iteration should not return any entries after segment 0 removal", 0, iterationCount.get());
+      } finally {
+         // Restore the original executor
+         TestingUtil.replaceField(originalExecutor, "executor", index, Index.class);
+      }
+   }
+}

--- a/core/src/test/java/org/infinispan/persistence/support/BatchAsyncStoreTest.java
+++ b/core/src/test/java/org/infinispan/persistence/support/BatchAsyncStoreTest.java
@@ -96,7 +96,9 @@ public class BatchAsyncStoreTest extends SingleCacheManagerTest {
    }
 
    @AfterClass(alwaysRun = true)
-   protected void clearTempDir() {
+   @Override
+   protected void destroyAfterClass() {
+      super.destroyAfterClass();
       Util.recursiveFileRemove(tmpDirectory);
    }
 }

--- a/tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java
+++ b/tools/src/main/java/org/infinispan/tools/store/migrator/file/SoftIndexFileStoreIterator.java
@@ -60,11 +60,11 @@ public class SoftIndexFileStoreIterator implements StoreIterator {
 
       SoftIndexIterator() {
          if (majorVersion < 11) {
-            this.fileProvider = new FileProvider(location, 1000, NonBlockingSoftIndexFileStore.PREFIX_10_1, 1024 * 1024);
+            this.fileProvider = new FileProvider(location, 1000, NonBlockingSoftIndexFileStore.PREFIX_10_1, 1024 * 1024, false);
             this.reader = EntryRecord::read10_1EntryHeader;
          } else {
             String prefix = majorVersion == 11 ? NonBlockingSoftIndexFileStore.PREFIX_11_0 : NonBlockingSoftIndexFileStore.PREFIX_12_0;
-            this.fileProvider = new FileProvider(location, 1000, prefix, 1024 * 1024);
+            this.fileProvider = new FileProvider(location, 1000, prefix, 1024 * 1024, false);
             this.reader = EntryRecord::readEntryHeader;
          }
          this.iterator = fileProvider.getFileIterator(null);

